pkg/ottl/ottlfuncs/func_merge.go

```diff
@@ -0,0 +1,85 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+const (
+	INSERT = "insert"
+	UPDATE = "update"
+	UPSERT = "upsert"
+)
+
+func Merge[K any](target ottl.Getter[K], source ottl.Getter[K], strategy string) (ottl.ExprFunc[K], error) {
+	if strategy != INSERT && strategy != UPDATE && strategy != UPSERT {
+		return nil, fmt.Errorf("invalid value for strategy, %v, must be 'insert', 'update' or 'upsert'", strategy)
+	}
+
+	return func(ctx context.Context, tCtx K) (interface{}, error) {
+		targetVal, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		if targetMap, ok := targetVal.(pcommon.Map); ok {
+			val, err := source.Get(ctx, tCtx)
+			if err != nil {
+				return nil, err
+			}
+
+			if valueMap, ok := val.(pcommon.Map); ok {
+				var mergeFunc func(k string, v pcommon.Value)
+				switch strategy {
+				case INSERT:
+					mergeFunc = func(k string, v pcommon.Value) {
+						if _, ok := targetMap.Get(k); !ok {
+							tv := targetMap.PutEmpty(k)
+							v.CopyTo(tv)
+						}
+					}
+				case UPDATE:
+					mergeFunc = func(k string, v pcommon.Value) {
+						if tv, ok := targetMap.Get(k); ok {
+							v.CopyTo(tv)
+						}
+					}
+				case UPSERT:
+					mergeFunc = func(k string, v pcommon.Value) {
+						if tv, ok := targetMap.Get(k); ok {
+							v.CopyTo(tv)
+						} else {
+							tv := targetMap.PutEmpty(k)
+							v.CopyTo(tv)
+						}
```

Comment:
don't need to do the get. PutEmpty is smart enough to not do a new allocation. And if it is not, then it should be :))

Reply:
@bogdandrutu adjusted.